### PR TITLE
Fail fast on creation and cleanup regardless of ctx - RunSnapshotRestore

### PR DESCRIPTION
**Pull request by KastenMike**
This PR does two things:
- Use a separate context from the main one for cleanup. This means if you have `WithTimeout(ctx, 5*time.Minute)` and the restore is too slow, it will still cleanup the cluster if you set the cleanup arg. Currently `.Delete` pod/pvc would fail on `context deadline exceeded` in such situations
- Adds an option for more fine grained timeouts on pod/pvc creation. Currently it will wait 15 minutes for the pod to start unless you set a shorter context timeout. This is quite high and likely if it will not start within 30-60 seconds it never will. 

Used a temp test to verify it works ok

Happy path:

image
image

Image pull / context.DeadlineExceeded - before the Cleanup wouldn't happen because of using the same context passed by the user (#102) -- assuming the passed context had less than 15 mins to timeout (e.g. k10tools):
image
image

### CODE REVIEW

File: pkg/csi/csi_ops.go [hunk 1]
```diff
@@ -83,11 +84,14 @@ func (o *validateOperations) ValidateVolumeSnapshotClass(ctx context.Context, vo
 type ApplicationCreator interface {
 	CreatePVC(ctx context.Context, args *types.CreatePVCArgs) (*v1.PersistentVolumeClaim, error)
 	CreatePod(ctx context.Context, args *types.CreatePodArgs) (*v1.Pod, error)
+	WaitForPVCReadyOrCheckEventIssues(ctx context.Context, namespace string, pvcName string) error
+	WaitForPodReadyOrCheckEventIssues(ctx context.Context, namespace string, podName string) error
 	WaitForPodReady(ctx context.Context, namespace string, podName string) error
 }
```
Comment (KastenMike) [inline pkg/csi/csi_ops.go:90]: @bathina2 or maybe better option than adding these 2 to the interface would be replacing `WaitForPodReady` content with calls to `waitForPVCReadyOrCheckEventIssues/Pod`...or add only `WaitForPVCReady` and replace the content of Pod with this new addition

**Review by ihcsim — COMMENTED** [review]
Thanks for working on this. 

Instead of adding a new `WaitForPodReadyOrCheckEventIssues()` function, what do you think of just wrapping the context for `kankube.WaitForPodReady()` with one that has a lower timeout value? The parent context's timeout value should override the default 15-min timeout. `WaitForPodReadyOrCheckEventIssues()` function feels a bit redundant.

The new `WaitForPVCReady()` function looks super useful, but I am not a big fan of having two different implementations of `poll.Wait()`. I am not sure I understand what's the issue with using the existing one. The `client rate limiter` error you saw was the client-side throttling warning, possibly indicating that the Kubernetes API server was under load, causing the context to timeout. If the existing one is not great, perhaps we should consider the ones from the [`k8s.io/apimachinery` package](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#UntilWithContext).

File: pkg/csi/csi_ops.go [hunk 2]
```diff
@@ -187,6 +191,107 @@ func (c *applicationCreate) CreatePod(ctx context.Context, args *types.CreatePod
...
+		FieldSelector: fmt.Sprintf("involvedObject.name=%s", name),
+	}
+
+	events, eventErr := c.kubeCli.CoreV1().Events(namespace).List(ctx, listOptions)
+	if eventErr != nil {
+		return errors.Wrapf(eventErr, "failed to retreieve events for %s of kind: %s", name, kind)
+	}
+
+	for _, event := range events.Items {
+		if event.Type == v1.EventTypeWarning {
```
[the remark this answers is not in the document]
Comment (ihcsim) [reply pkg/csi/csi_ops.go:288]: Ah.. nvm; I just saw the field selector :+1:.

File: pkg/csi/csi_ops.go [hunk 3]
```diff
@@ -187,6 +191,107 @@ func (c *applicationCreate) CreatePod(ctx context.Context, args *types.CreatePod
...
+		pvc, err := c.kubeCli.CoreV1().PersistentVolumeClaims(namespace).Get(timeoutCtx, name, metav1.GetOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "Could not find PVC")
+		}
+
+		if pvc.Status.Phase == v1.ClaimLost {
+			return fmt.Errorf("failed to create a PVC, ClaimLost")
+		}
+
+		if pvc.Status.Phase != "" && pvc.Status.Phase != v1.ClaimPending {
```
[the remark this answers is not in the document]
Comment (KastenMike) [reply pkg/csi/csi_ops.go:222]: I guess it could be if the only invalid option is ClaimLost

File: pkg/csi/csi_ops.go [hunk 4]
```diff
@@ -187,6 +191,107 @@ func (c *applicationCreate) CreatePod(ctx context.Context, args *types.CreatePod
...
+		}
+
+		if pod.Status.Phase == v1.PodPending {
+			if pod.Status.Reason == "OutOfmemory" || pod.Status.Reason == "OutOfcpu" {
+				return errors.Errorf("Pod stuck in pending state, reason: %s", pod.Status.Reason)
+			}
+		}
+
+		if pod.Status.Phase != "" && pod.Status.Phase != v1.PodPending {
+			return nil
```
[the remark this answers is not in the document]
Comment (KastenMike) [reply pkg/csi/csi_ops.go:265]: hmm ok...the `WaitForPodReady`  used has this approach

File: pkg/csi/csi_ops.go [hunk 5]
```diff
@@ -187,6 +191,107 @@ func (c *applicationCreate) CreatePod(ctx context.Context, args *types.CreatePod
...
+			return context.DeadlineExceeded
+		default:
+			time.Sleep(basicK8sObjectWait)
+		}
+	}
+}
+
+func (c *applicationCreate) getErrorFromEvents(ctx context.Context, namespace, name, kind string) error {
+	listOptions := metav1.ListOptions{
+		TypeMeta:      metav1.TypeMeta{Kind: kind},
```
[the remark this answers is not in the document]
Comment (KastenMike) [reply pkg/csi/csi_ops.go:278]: No, this is gathering events related to a particular pvc/pod. e.g. here kind might be "Pod" and `involvedObject.name=my-kannisterpod` and it gathers events matching these filters

File: pkg/csi/csi_ops.go [hunk 6]
```diff
@@ -187,6 +191,107 @@ func (c *applicationCreate) CreatePod(ctx context.Context, args *types.CreatePod
...
+	}
+
+	events, eventErr := c.kubeCli.CoreV1().Events(namespace).List(ctx, listOptions)
+	if eventErr != nil {
+		return errors.Wrapf(eventErr, "failed to retreieve events for %s of kind: %s", name, kind)
+	}
+
+	for _, event := range events.Items {
+		if event.Type == v1.EventTypeWarning {
+			fmt.Errorf(event.Message)
```
[the remark this answers is not in the document]
Comment (KastenMike) [reply pkg/csi/csi_ops.go:289]: ah copy/paste error :D

File: pkg/csi/csi_ops.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ihcsim) [reply pkg/csi/csi_ops.go:278]: I thought the `Pod` kind would be part of your field selector. E.g., see https://github.com/kubernetes/kubernetes/blob/e844c12a6184433ea510b39004d579f86e7411b2/test/e2e/storage/csi_mock_volume.go#L2022-L2030.

[the remark this answers is not in the document]
Comment (KastenMike) [reply pkg/csi/csi_ops.go:278]: ah ok. Not sure if they are different. Can update to the other

**Comment by KastenMike** [comment]
> Thanks for working on this.
> 
> Instead of adding a new `WaitForPodReadyOrCheckEventIssues()` function, what do you think of just wrapping the context for `kankube.WaitForPodReady()` with one that has a lower timeout value? The parent context's timeout value should override the default 15-min timeout. `WaitForPodReadyOrCheckEventIssues()` function feels a bit redundant.
> 
> The new `WaitForPVCReady()` function looks super useful, but I am not a big fan of having two different implementations of `poll.Wait()`. I am not sure I understand what's the issue with using the existing one. The `client rate limiter` error you saw was the client-side throttling warning, possibly indicating that the Kubernetes API server was under load, causing the context to timeout. If the existing one is not great, perhaps we should consider the ones from the [`k8s.io/apimachinery` package](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#UntilWithContext).

I can look into it again but in k10 it seemed unreliable to me (at least with unit tests). `errors.Is` didn't always catch poll.Wait errors for ctx DeadlineExceeded or Cancelled with the rate limit error so I switched to this case. Originally k10 also used the pod wait for ready and was switched because of this dodgy rate limiter complaints

**Comment by ihcsim** [comment]
@KastenMike I think that might be because the client-side throttling error [wraps](https://github.com/kubernetes/kubernetes/blob/4720f0725c3dadbfec5d90b6464d3535a62692cf/staging/src/k8s.io/client-go/rest/request.go#L519-L521) the context expired error. k10 might need to use `errors.As`.

**Comment by KastenMike** [comment]
> @KastenMike I think that might be because the client-side throttling error [wraps](https://github.com/kubernetes/kubernetes/blob/4720f0725c3dadbfec5d90b6464d3535a62692cf/staging/src/k8s.io/client-go/rest/request.go#L519-L521) the context expired error. k10 might need to use `errors.As`.

@ihcsim it is already using `erorrs.Is(err, context.DeadlineExceeded)`

File: pkg/csi/csi_ops.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (KastenMike) [reply pkg/csi/csi_ops.go:265]: I'll remove this and move the events reading code to WaitForPodyReady with a shorter timeout as you suggested

**Comment by ihcsim** [comment]
@KastenMike So sounds like the issue isn't with the `poll.Wait()` function, but that we are not comparing the right things. `errors.Is` matches `err` and `target` by value, not by their error strings. Also, I was hoping we can use `errors.As()`, not `errors.Is()`, but it won't help in this case.

Any reasons why we need to single out `context.DeadlineExceeded`? Can we just retrieve the pod's event errors (if any) as long as `if err != nil`? Maybe except for `context.Canceled`, as that usually indicates a request from the caller to cancel the current operation.

**Comment by KastenMike** [comment]
> Can we just retrieve the pod's event errors (if any) as long as `if err != nil`? 

Can try

File: pkg/csi/snapshot_restore.go [hunk 7]
```diff
@@ -261,44 +278,50 @@ func (s *snapshotRestoreSteps) RestoreApplication(ctx context.Context, args *typ
...
+	}
+
 	if err = s.createAppOps.WaitForPodReady(ctx, args.Namespace, pod.Name); err != nil {
 		return pod, pvc, errors.Wrap(err, "Pod failed to become ready")
 	}
 	return pod, pvc, nil
 }
 
-func (s *snapshotRestoreSteps) Cleanup(ctx context.Context, results *types.CSISnapshotRestoreResults) {
+func (s *snapshotRestoreSteps) Cleanup(results *types.CSISnapshotRestoreResults) {
```
Comment (bathina2) [inline pkg/csi/snapshot_restore.go:292]: Why are you changing the function signature by removing the context?

Comment (bathina2) [reply pkg/csi/snapshot_restore.go:292]: The context is being used.

File: pkg/csi/csi_ops_test.go [hunk 8]
```diff
@@ -482,17 +484,7 @@ func (s *CSITestSuite) TestCreatePod(c *C) {
 			podChecker: NotNil,
 		},
 		{
-			cli: fake.NewSimpleClientset(),
```
Comment (bathina2) [inline pkg/csi/csi_ops_test.go:485]: Why did you remove this valid test case?

File: pkg/csi/snapshot_restore.go (hunk 7 above)
Comment (bathina2) [reply pkg/csi/snapshot_restore.go:292]: How does this help "Fail fast"

File: pkg/csi/snapshot_restore.go [hunk 9]
```diff
@@ -24,6 +24,11 @@ const (
 	createdByLabel          = "created-by-kubestr-csi"
 	clonePrefix             = "kubestr-clone-"
 	snapshotPrefix          = "kubestr-snapshot-"
+
+	basicK8sObjectWait = 1 * time.Second // Sleeps to avoid burning resources on kubectl calls and to make testing more reliable
+
+	PVCKind = "PersistentVolumeClaim"
```
Comment (bathina2) [inline pkg/csi/snapshot_restore.go:30]: If these are being used elsewhere, define them there and not here.

File: pkg/csi/snapshot_restore.go (hunk 7 above)
Comment (KastenMike) [reply pkg/csi/snapshot_restore.go:292]: It doesn't help "fail fast"
It refers to the "and cleanup regardless". It also prevents https://github.com/kastenhq/kubestr/issues/102

Comment (KastenMike) [reply pkg/csi/snapshot_restore.go:292]: When PVC/Pod hangs the passed ctx gets DeadlineExceeded  and cleanup never happens. This avoids that issues

File: pkg/csi/csi_ops_test.go (hunk 8 above)
Comment (bathina2) [reply pkg/csi/csi_ops_test.go:485]: got it.

File: pkg/csi/snapshot_restore.go (hunk 7 above)
Comment (KastenMike) [reply pkg/csi/snapshot_restore.go:292]: unless someone outside this repo might be using this API directly (not sure how likely)...then could basically keep the passed context but just ignore it although it looks kinder weird

File: pkg/csi/snapshot_restore.go (hunk 9 above)
Comment (KastenMike) [reply pkg/csi/snapshot_restore.go:30]: I moved these 2. Some of the existing const's are also only used in `csi_ops` but I just left them here

File: pkg/csi/snapshot_restore.go (hunk 7 above)
Comment (KastenMike) [reply pkg/csi/snapshot_restore.go]: I changed the timeout to 2 mins for the PVC and then the Pod being ready (so 4 mins total worst case). There's no need to have a timeout on kubectl commands; the hanging is all in the PVCs/Pods reaching ready state. The cleanup here just requests deletion, it doesn't wait around for deletion to complete (unlike I do in k10 to allow 1 retry of creation for a specific scenario [`'resource.sizeGb': '1'. Disk size cannot be smaller than 200 GB`])

Comment (ihcsim) [reply pkg/csi/snapshot_restore.go]: Yes, you are right. Delete does return immediately, even though the underlying resources are stuck in the `Terminating` state, which is a separate problem.

File: pkg/csi/csi_ops.go [hunk 10]
```diff
@@ -187,14 +199,92 @@ func (c *applicationCreate) CreatePod(ctx context.Context, args *types.CreatePod
 	return podRes, nil
 }
 
+func (c *applicationCreate) WaitForPVCReady(ctx context.Context, namespace, name string) error {
+	if c.kubeCli == nil {
+		return fmt.Errorf("kubeCli not initialized")
+	}
+
+	err := c.waitForPVCReady(ctx, namespace, name)
+	if err != nil && (errors.Is(err, context.DeadlineExceeded) || strings.Contains(err.Error(), "would exceed context deadline")) {
```
Comment (ihcsim) [inline pkg/csi/csi_ops.go:208]: Can we just do `if err != nil` here? I don't see why we need to single out `context.DeadlineExceeded`, nor the string comparison.

File: pkg/csi/csi_ops.go [hunk 11]
```diff
@@ -187,14 +199,92 @@ func (c *applicationCreate) CreatePod(ctx context.Context, args *types.CreatePod
...
+	})
+}
+
 func (c *applicationCreate) WaitForPodReady(ctx context.Context, namespace string, podName string) error {
 	if c.kubeCli == nil {
 		return fmt.Errorf("kubeCli not initialized")
 	}
-	err := kankube.WaitForPodReady(ctx, c.kubeCli, namespace, podName)
+	err := c.waitForPodReady(ctx, namespace, podName)
+	if err != nil && (errors.Is(err, context.DeadlineExceeded) || strings.Contains(err.Error(), "would exceed context deadline")) {
```
Comment (ihcsim) [inline pkg/csi/csi_ops.go:244]: As above, can we just do `if err != nil` here? I don't see why we need to single out `context.DeadlineExceeded`, nor the string comparison.

File: pkg/csi/csi_ops_test.go [hunk 12]
```diff
@@ -1160,3 +1159,110 @@ func (s *CSITestSuite) TestDeleteSnapshot(c *C) {
...
+	normalGetFunc :=  func(action k8stesting.Action) (handled bool, ret runtime.Object, err error) {
+		return
+	}
+	deadlineExceededGetFunc := func(action k8stesting.Action) (handled bool, ret runtime.Object, err error) {
+		return true, nil, pkgerrors.Wrapf(context.DeadlineExceeded, "some wrapped error")
+	}
+	// side effect of poll.Wait. See https://github.com/kubernetes/kubernetes/blob/42fec42586c4411f649bf766fe65059602ee6499/vendor/golang.org/x/time/rate/rate.go#L247
+	preemptiveDeadlineExceededGetFunc := func(action k8stesting.Action) (handled bool, ret runtime.Object, err error) {
+		return true, nil, pkgerrors.Wrapf(fmt.Errorf("rate: Wait(n=1) would exceed context deadline"), "some wrapped error")
+	}
```
Comment (ihcsim) [inline pkg/csi/csi_ops_test.go:1179]: Just feels like the source and test code will be simpler if we just always check for and retrieve the warning events regardless of what the error kinds are.

**Review by ihcsim — COMMENTED** [review]
Looks good. Also, thanks for the unit tests. One last item, as previously mentioned, why do we need to single out the handling of the `context.DeadlineExceeded` error? Why not just always grab the warning events (if any), when `if err != nil`?

**Comment by KastenMike** [comment]
> Looks good. Also, thanks for the unit tests. One last item, as previously mentioned, why do we need to single out the handling of the `context.DeadlineExceeded` error? Why not just always grab the warning events (if any), when `if err != nil`?

I forgot to address it. Will make the update Tomorrow

File: pkg/csi/csi_ops.go (hunk 11 above)
Comment (ihcsim) [reply pkg/csi/csi_ops.go:244]: Yeah, or just wrap the warning events as string (if any) with the original event.

**Comment by KastenMike** [comment]
> Why not just always grab the warning events (if any), when `if err != nil`?

done

**Review by ihcsim — APPROVED** [review]
LGTM. Thanks again for working on this :+1:

**Review by bathina2 — APPROVED** [review]
